pkg/oc/cli/cmd/push/binary.go

```diff
@@ -0,0 +1,368 @@
+		if err != nil {
+			return fmt.Errorf("--from requires a valid image stream tag or image reference: %v", err)
+		}
+		if len(ref.Registry) > 0 || (len(ref.ID) == 0 && len(ref.Tag) == 0) {
+			return fmt.Errorf("--from must be [NAMESPACE/]NAME:TAG or [NAMESPACE/]NAME@ID")
+		}
+	}
+	return nil
+}
+
+func (o *PushBinaryOptions) Run() error {
+	localPath, r, size, err := contentsForPathOrURL(o.Path, o.In)
+	if err != nil {
+		return err
+	}
+	r = convertStreamToTarGz(localPath, r, size)
+
+	to, err := imageapi.ParseDockerImageReference(o.To)
+	if err != nil {
+		return fmt.Errorf("--to is not a valid reference: %v", err)
+	}
+	if len(to.Namespace) == 0 {
+		to.Namespace = o.Namespace
+	}
+
+	imageMetadata := &docker10.DockerImage{
+		Config: &docker10.DockerConfig{
+			Labels: map[string]string{
+				"io.openshift.image.binary": "true",
+			},
+		},
+	}
+	if o.Time {
+		imageMetadata.Created = metav1.Now()
+	}
+	clone := &v1.ImageStreamTagClone{
+		ObjectMeta: metav1.ObjectMeta{Name: to.NameString(), Namespace: to.Namespace},
+		Image: &v1.ImageCloneMetadata{
+			DockerImageMetadata: runtime.RawExtension{
+				Object: imageMetadata,
+			},
+		},
+	}
+	if len(o.BaseImage) > 0 {
+		ref, err := imageapi.ParseDockerImageReference(o.BaseImage)
+		if err != nil {
+			return fmt.Errorf("--from requires a valid image stream tag or image reference: %v", err)
+		}
+		clone.From = &kapiv1.ObjectReference{Name: ref.NameString(), Namespace: ref.Namespace}
+		if len(ref.ID) > 0 {
+			clone.From.Kind = "ImageStreamImage"
+		} else {
+			clone.From.Kind = "ImageStreamTag"
+		}
+	}
+
+	result, err := o.Client.ImageStreams(to.Namespace).CloneWithLayer(clone, r)
+	if err != nil {
+		if errors.IsAlreadyExists(err) {
+			glog.V(2).Infof("Unable to push: %v", err)
+			return fmt.Errorf("a push has already been requested for this tag, use --force to push anyway")
+		}
+		return fmt.Errorf("failed to push binary: %v", err)
+	}
+	obj, err := kapi.Scheme.ConvertToVersion(result, imageapi.SchemeGroupVersion)
+	if err != nil {
+		return fmt.Errorf("unable to convert: %v", err)
+	}
+	info, err := o.Mapper.InfoForObject(obj, nil)
+	if err != nil {
+		return err
+	}
+	return o.PrintObject([]*resource.Info{info})
+}
+
+func contentsForPathOrURL(s string, in io.Reader, subpaths ...string) (string, io.Reader, int64, error) {
+	switch {
+	case s == "-":
+		return "", in, -1, nil
+	case strings.Index(s, "http://") == 0 || strings.Index(s, "https://") == 0:
+		_, err := url.Parse(s)
+		if err != nil {
+			return "", nil, 0, fmt.Errorf("the URL passed to filename %q is not valid: %v", s, err)
+		}
+		res, err := http.Get(s)
+		if err != nil {
+			return "", nil, 0, err
+		}
+		return "", res.Body, -1, nil
+	default:
+		stat, err := os.Stat(s)
+		if err != nil {
+			return "", nil, 0, err
+		}
+		if !stat.IsDir() {
+			contents, err := os.Open(s)
+			return s, contents, stat.Size(), err
+		}
+		for _, sub := range subpaths {
```

Comment:
Picking up the first non-directory entry in a given directory seems wrong. The user should be informed that he gave a directory instead of an expected file.